controllers/podiochaos/controller.go

```diff
@@ -0,0 +1,161 @@
+// Copyright 2020 Chaos Mesh Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package podiochaos
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	"github.com/chaos-mesh/chaos-mesh/controllers/utils/chaosdaemon"
+
+	"github.com/go-logr/logr"
+	v1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/client-go/util/retry"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/chaos-mesh/chaos-mesh/api/v1alpha1"
+	"github.com/chaos-mesh/chaos-mesh/pkg/chaosdaemon/pb"
+)
+
+// Reconciler applys podioworkchaos
+type Reconciler struct {
+	client.Client
+	client.Reader
+	Recorder record.EventRecorder
+
+	Log logr.Logger
+}
+
+func (r *Reconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
+	ctx := context.TODO()
+
+	obj := &v1alpha1.PodIoChaos{}
+
+	if err := r.Client.Get(ctx, req.NamespacedName, obj); err != nil {
+		if apierrors.IsNotFound(err) {
+			r.Log.Info("chaos not found")
+		} else {
+			// TODO: handle this error
+			r.Log.Error(err, "unable to get chaos")
+		}
+		return ctrl.Result{}, nil
+	}
+
+	r.Log.Info("updating io chaos", "pod", obj.Namespace+"/"+obj.Name, "spec", obj.Spec)
+
+	pod := &v1.Pod{}
+
+	err := r.Client.Get(ctx, types.NamespacedName{
+		Name:      obj.Name,
+		Namespace: obj.Namespace,
+	}, pod)
+	if err != nil {
+		r.Log.Error(err, "fail to find pod")
+		return ctrl.Result{}, nil
+	}
+
+	failedMessage := ""
+	observedGeneration := obj.ObjectMeta.Generation
+	pid := obj.Status.Pid
+	startTime := obj.Status.StartTime
+	defer func() {
+		if err != nil {
+			failedMessage = err.Error()
+		}
+
+		updateError := retry.RetryOnConflict(retry.DefaultBackoff, func() error {
+			obj := &v1alpha1.PodIoChaos{}
+
+			if err := r.Client.Get(context.TODO(), req.NamespacedName, obj); err != nil {
+				r.Log.Error(err, "unable to get chaos")
+				return err
+			}
+
+			obj.Status.FailedMessage = failedMessage
+			obj.Status.ObservedGeneration = observedGeneration
```

Comment:
It seems the `failedMessage`, `observedGeneration`, `pid` and `startTime` will never change.

Reply:
`observedGeneration` is initialized with the newest generation, and `pid`, `startTime` are modified in the following part.

`failedMessage` is modified in line 80.

Reply:
Yes, I see. But in function RetryOnConflict, the newest generation may change.